param/__init__.py

```diff
@@ -1461,8 +1488,15 @@ def serialize(cls, value):
 
     @classmethod
     def deserialize(cls, value):
-        from numpy import asarray
-        return asarray(value)
+        import numpy
+        try:
+            return _deserialize_from_path(
+                {'.npy': numpy.load, '.txt': lambda x: numpy.loadtxt(str(x))},
+                value, 'Array'
+            )
+        except:
+            pass
+        return numpy.asarray(value)
```

Comment:
So, what happens if numpy.load raises an exception and is unable to parse it? Will the user get any error messages? Seems like they would get an obscure message here about value not being suitable for numpy.asarray.

Reply:
They will be warned with the "Could not parse file {} as {}" bit.

Reply:
It also prints the stack trace viz. "exc_info=True", which includes the original error.

Reply:
> Ok, thanks. I think I see that better now. However, I can't seem to trigger a meaningful error message like that:
> 
> Somehow it seems like there is an uncaught exception rather than a logging message, but I can't spot where that is.

Ah, that's because we don't even try to deserialize if we can't find the disk on file - it just falls through to the previous behaviour.

```
(D:\conda-envs\param) PS D:\Documents\Projects\Code\param> echo "fiiii" > a.pkl
(D:\conda-envs\param) PS D:\Documents\Projects\Code\param> python -c "import param; param.DataFrame.deserialize('a.pkl')"
WARNING:param: Could not parse file 'a.pkl' as DataFrame
Traceback (most recent call last):
  File "D:\Documents\Projects\Code\param\param\__init__.py", line 304, in _deserialize_from_path
    return ext_to_routine[ext](path)
  File "D:\conda-envs\param\lib\site-packages\pandas\io\pickle.py", line 203, in read_pickle
    return pickle.load(handles.handle)  # type: ignore[arg-type]
_pickle.UnpicklingError: invalid load key, '\xff'.
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "D:\Documents\Projects\Code\param\param\__init__.py", line 1600, in deserialize
    return pandas.DataFrame(value)
  File "D:\conda-envs\param\lib\site-packages\pandas\core\frame.py", line 590, in __init__
    raise ValueError("DataFrame constructor not properly called!")
ValueError: DataFrame constructor not properly called!
```

> In any case, it would be nice to eliminate the two levels of exception handling that make the code verbose and difficult to follow. Seems like this method could be simply be:
> 
> ```
>     def deserialize(cls, value):
>         return numpy.asarray(_deserialize_from_path(
>             {'.npy': numpy.load, '.txt': lambda x: numpy.loadtxt(str(x))},
>             value, 'Array'))
> ```
> 
> and either we allow the underlying method to raise a useful exception, or else we have it print a warning and then return a dummy value. I don't see what's gained by the extra try/except, but could easily be missing something.

I was trying to be as conservative as possible with the assumptions: "is it a string? a path to an existing file? ends in such and such extension? Did we successfully parse said file?" If the answer is no to any of these we fall back at the cost of more cryptic failures.

Looking again, I'm probably being too paranoid about it. "Is it a string?" is probably enough. I think the only possible valid array we're clobbering is a scalar numpy string array

``` python
>>> import numpy as np
>>> np.asarray("ab")
array('ab', dtype='<U2')
```
Which I'd imagine would be a rare intent.

Reply:
Yes, I think we can explicitly consider such an array not a valid final value for the Parameter.  Given that in any application where filenames and paths are being used an incorrect or outdated path is the most likely error by far, I think we should optimize the error reporting around (a) indicating that we are treating the provided string as a filename and (b) indicating that the file is not found. In that circumstance, there should be a simple, straightforward error message.  All other cases can be lumped together a bit, hopefully being clear but not needing to be simple or tightly controlled like that. Parse errors on pickle files do happen, but they are very rare compared to not finding the file at all!

Reply:
Ok, I've updated the numpy code to the below (pandas is similar):

``` python
    @classmethod
    def deserialize(cls, value):
        import numpy
        if isinstance(value, unicode):
            return _deserialize_from_path(
                {'.npy': numpy.load, '.txt': lambda x: numpy.loadtxt(str(x))},
                value, 'Array'
            )
        else:
            return numpy.asarray(value)
```

`_deserialize_from_path` doesn't return if it can't deserialize. This is as per the discussion above: if you've got a string and it's not a path, it's an error.